diff --git a/include/mbedtls/ssl.h b/include/mbedtls/ssl.h
index 209dbf605..e7b4f717b 100644
--- a/include/mbedtls/ssl.h
+++ b/include/mbedtls/ssl.h
@@ -1466,6 +1466,7 @@ struct mbedtls_ssl_context
                             *   Possible values are #MBEDTLS_SSL_CID_ENABLED
                             *   and #MBEDTLS_SSL_CID_DISABLED. */
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
+    void *appData;
 };

 #if defined(MBEDTLS_SSL_HW_RECORD_ACCEL)
