diff --git a/include/mbedtls/ssl.h b/include/mbedtls/ssl.h
index 209dbf605..e7b4f717b 100644
--- a/include/mbedtls/ssl.h
+++ b/include/mbedtls/ssl.h
@@ -1466,6 +1466,7 @@ struct mbedtls_ssl_context
                             *   Possible values are #MBEDTLS_SSL_CID_ENABLED
                             *   and #MBEDTLS_SSL_CID_DISABLED. */
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
+    void *appData;
 };

 #if defined(MBEDTLS_SSL_HW_RECORD_ACCEL)
diff --git a/library/sha512.c b/library/sha512.c
index 02a135ca9..020842cba 100644
--- a/library/sha512.c
+++ b/library/sha512.c
@@ -153,6 +153,7 @@ void mbedtls_sha512_starts( mbedtls_sha512_context *ctx,
 /*
  * Round constants
  */
+#define K KK
 static const uint64_t K[80] =
 {
     UL64(0x428A2F98D728AE22),  UL64(0x7137449123EF65CD),
