diff --git a/include/mbedtls/ssl.h b/include/mbedtls/ssl.h
index dffc16219..5ea1cc165 100644
--- a/include/mbedtls/ssl.h
+++ b/include/mbedtls/ssl.h
@@ -916,6 +916,7 @@ struct mbedtls_ssl_context
     char own_verify_data[MBEDTLS_SSL_VERIFY_DATA_MAX_LEN]; /*!<  previous handshake verify data */
     char peer_verify_data[MBEDTLS_SSL_VERIFY_DATA_MAX_LEN]; /*!<  previous handshake verify data */
 #endif
+    void *appData;
 };

 #if defined(MBEDTLS_SSL_HW_RECORD_ACCEL)
